enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,247 @@
+  - each certificate may contain:
+    - Single Common Name containing The API server DNS/IPAddress or a wildcard entry (wildcard certificate).
+    - Multiple Subject Alternative Names (SAN) containing the API server DNS/IPAddress.
+
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate.
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+Automatic certificate rotation and integration with 3rd party cert issuing services.
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file.
+
+A new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+For each provided certificate, the following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+Certificate files will be read from their configured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+because we dont want to disrupt the internal API communication and make sure the internal certificate will be automaticly renewed,
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+
+### Kubeconfig Generation
+Each configured certificate can contain multiple FQDN and wildcards values, for each unique FQDN address kubeconfig file will be generated on the filesystem.
+
+Every generated `kubeconfig` for the custom certificates will omit the certificate-authority-data section,
+therefore custom certificates will have to be validated against CAs in the RHEL Client trust store. 
+
+each certificate will be examined and deteremined if its wildcard only.
+certificate will be considered as a wildcard only if there are no FQDN Entries found.
+the FQDN Address is searched in:
+- Certificate CN
+- Certificate Subject Alternative Name (SAN)
+- names configuration value.
+
+when a Certificate found as a wildcard only , host public IP Address will be placed inside its kubeconfig's `server` section.
+
+Certain values that configured in `names` field can cause  internal API communication issues  ie:127.0.0.1,localhost therefore we must not allow them.
+
+when the `names` field is not contained in the certificate (ie: included in the wildcard),
```

Comment:
Not sure I understand this section. If one of the `names` fields is not contained in the CN/SAN, then MicroShift should include it, right?
Or do you mean the `names` was not configured? If that was the case then the example below should not happen, as the kubeconfig is generated from available SANs (be those through configuration or default behavior).

Reply:
this section actually doesnt belong to kubeconfig generation,
its more  a `Support Procedures` ,  the problem  with `names` configuration that it doesn't replace the SNI inside certificate that is used for TLS  validation. its more  extension to  the certificates.

Reply:
@dhellmann ,  suggested we dont perform any clever validations here, since its the way it works in OCP.

Reply:
If its an extension then is it automatically added to the names in the cert? If so the first case I described can not happen. If it is already included it should not hurt, right?